### ISSUE TITLE
Plugins - Question

### ISSUE THREAD

**Issue by MoroseCorpse** [Splamy/TS3AudioBot#695]
I have a general question about bot plugins, because I noticed that the bots always do their tasks one after the other and never in parallel. 
Is that true or am I wrong?

If it's true, is it possible that the bots can always run events and other things in parallel?

**Comment by Splamy** [comment]
Can you give more details? What is a 'task' for you? If you mean on call like `!x (!y (!z))` then yes, those are always sequentially per bot.
I don't think those need to be async, this only makes stuff more complicated, more error prone, and has little benefits.

If you need parallelism/asynchronity in your plugin for whatever reason, you can always use c#'s language features like `Task`s and `Thread`s, as well as the `TickPool` prvided by the audiobot, which acts as a single threaded timer (however I want to rework the TickPool a bit soon)

**Comment by MoroseCorpse** [comment]
I meant whether the bot is running parallel events.

For example, if the clientmoved event is executed and this code is a bit longer in it or it hangs up somehow, that other events or the same can still be executed in parallel.

This would be mega good, because at the moment I can't find the error in my plugin, why it hangs sometimes and then everything would still work and only sometimes it takes some users a bit longer.

P.S: Sorry for the late answer. Had stopped programming for a while

**Comment by Splamy** [comment]
The event dispatcher is a single threaded event loop; this means all events are processed sequentially.
Usually you should not have task that block that long.
If you need to execute something asynchronous you can move that out to a separate Task
For example:
```cs
// register just like normal
ts3FullClient.OnChannelCreated += Ts3FullClient_OnChannelCreated;

private void Ts3FullClient_OnChannelCreated(object sender, IEnumerable<ChannelCreated> e)
{
	Task.Run(() =>
	{
		// Do long running work within this task
		// ...
	}).ConfigureAwait(false);
}
```
